controllers/periodic_healthcheck.go

```diff
@@ -0,0 +1,157 @@
+					if err != nil {
+						r.Log.Error(err, "Error filtering machines for etcd cluster")
+					}
+
+					ownedMachines := etcdMachines.Filter(collections.OwnedMachines(&ec))
+					endpointToMachineMapper := make(map[string]*clusterv1.Machine)
+					for _, m := range ownedMachines {
+						machineClientURL := getMemberClientURL(getEtcdMachineAddress(m))
+						endpointToMachineMapper[machineClientURL] = m
+					}
+
+					r.etcdHealthCheckConfig.etcdadmClusterMapper[ec.UID] = etcdadmClusterMemberHealthConfig{
+						unhealthyMembersFrequency: make(map[string]int),
+						unhealthyMembersToRemove:  make(map[string]*clusterv1.Machine),
+						endpointToMachineMapper:   endpointToMachineMapper,
+						cluster:                   cluster,
+						ownedMachines:             ownedMachines,
+					}
+				} else {
+					cluster = r.etcdHealthCheckConfig.etcdadmClusterMapper[ec.UID].cluster
+				}
+
+				if err := r.periodicEtcdMembersHealthCheck(ctx, cluster, &ec); err != nil {
+					r.Log.Error(err, "Error performing healthcheck")
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (r *EtcdadmClusterReconciler) periodicEtcdMembersHealthCheck(ctx context.Context, cluster *clusterv1.Cluster, etcdCluster *etcdv1.EtcdadmCluster) error {
+	currClusterHFConfig := r.etcdHealthCheckConfig.etcdadmClusterMapper[etcdCluster.UID]
+	for _, endpoint := range strings.Split(etcdCluster.Status.Endpoints, ",") {
+		err := r.performEndpointHealthCheck(ctx, cluster, endpoint, true)
+		if err != nil {
+			// member failed healthcheck so add it to unhealthy map or update it's unhealthy count
+			currClusterHFConfig.unhealthyMembersFrequency[endpoint]++
+			r.Log.Info(fmt.Sprintf("Member %v failed healthcheck, adding it to unhealthyMembers list", endpoint))
+			if currClusterHFConfig.unhealthyMembersFrequency[endpoint] >= maxUnhealthyCount {
+				r.Log.Info(fmt.Sprintf("Member %v failed healthcheck %d times, adding it to list of members to remove", endpoint, maxUnhealthyCount))
+				// member has been unresponsive, add the machine to unhealthyMembersToRemove queue
+				m := currClusterHFConfig.endpointToMachineMapper[endpoint]
+				currClusterHFConfig.unhealthyMembersToRemove[endpoint] = m
+			}
+		} else {
+			// member passed healthcheck. so if it was previously added to unhealthy map, remove it since only consecutive failures should lead to member removal
+			if _, markedUnhealthy := currClusterHFConfig.unhealthyMembersFrequency[endpoint]; markedUnhealthy {
+				delete(currClusterHFConfig.unhealthyMembersFrequency, endpoint)
+			}
+		}
+	}
+
+	if len(currClusterHFConfig.unhealthyMembersToRemove) > 0 {
+		ep, err := NewEtcdPlane(ctx, r.Client, currClusterHFConfig.cluster, etcdCluster, currClusterHFConfig.ownedMachines)
+		if err != nil {
+			return errors.Wrap(err, "Error initializing internal object EtcdPlane")
+		}
+		// delete machine
+		finalEndpoints := make(map[string]bool)
```

Comment:
lines 134-137 and line 147-150 are mainly to get the updated list of endpoints removing the ones that have failed healthcheck. Maybe there's a better way of doing this :)